paddle/fluid/pybind/tensor_py.h

```diff
@@ -64,6 +70,22 @@ namespace pybind {
 
 namespace details {
 
+class PYBIND11_HIDDEN NumpyAllocation : public memory::Allocation {
+ public:
+  explicit NumpyAllocation(PyObject *arr, size_t size,
```

Comment:
The design of `NumpyAllocation` can be improved. There are a few problems in the current implementation:

- There is no mechanism to check whether `PyObject *arr` is exactly an object of `PyArrayObject`. You use `reinterpret_cast` below, but it is not type safe.
- There is not necessary to parse `arr` and `size` in the same time. Information of `size` can be calculated from `arr` itself.
- Increase reference count outside `NumpyAllocation` and decrease reference count inside `NumpyAllocation` is not good.
- Check the `place` is `CPUPlace` exactly or just remove this parameter.
- `explicit` can be removed.

Reply:
done